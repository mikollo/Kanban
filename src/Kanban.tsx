import React, { useReducer } from "react";
import List from "./components/List";
import ListCreator from "./components/ListCreator";

export default function Kanban() {
  const [lists, dispatch] = useReducer(kanbanReducer, []);
  return (
    <div>
      {JSON.stringify(lists)}
      {/*  {lists.map((data, index) => (
        <List indexOfList={index} tasks={data} dispatch={dispatch} />
      ))} */}
      <ListCreator dispatch={dispatch} />
    </div>
  );
}

function kanbanReducer(state: KanbanInterface, action: actionInterface) {
  let newState = state;
  switch (action.type) {
    case "add list":
      if (action.newListName)
        newState.push({ name: action.newListName, tasks: [] });
      break;
    case "add task to list":
      if (action.indexOfList && action.newTaskData)
        newState[action.indexOfList].tasks.push(action.newTaskData);
      break;
    case "change list name":
      if (action.indexOfList && action.newListName)
        newState[action.indexOfList] = {
          name: action.newListName,
          tasks: newState[action.indexOfList].tasks
        };
      break;
    case "edit task":
      if (action.indexOfList && action.indexOfTask && action.newTaskData)
        newState[action.indexOfList].tasks[action.indexOfTask] =
          action.newTaskData;
      break;
    default:
      throw "Action doesn't have a valid type";
  }
  return newState;
}

export interface actionInterface {
  type: "add list" | "add task to list" | "change list name" | "edit task";
  newListName?: string;
  newTaskData?: TaskInterface;
  indexOfList?: number;
  indexOfTask?: number;
}

export type KanbanInterface = ListInterface[];

export interface ListInterface {
  name: string;
  tasks: TaskInterface[];
}

export interface TaskInterface {
  title: string;
  description: string;
}
