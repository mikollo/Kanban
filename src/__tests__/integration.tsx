import React from "react";
import { render, fireEvent, cleanup } from "react-testing-library";
// this adds custom jest matchers from jest-dom
import "jest-dom/extend-expect";
import Kanban from "../Kanban";

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

test("List can be added", async () => {
  const { getByText, getByPlaceholderText, container } = render(<Kanban />);
  const input = getByPlaceholderText("New list name");
  (input as any).value = "Some list";
  fireEvent.change(input);
  fireEvent.click(getByText("Add new list"));
  const createdListHeader = getByText("Add new task");
  expect(createdListHeader).toBeDefined();
  expect(container).toMatchSnapshot();
});
