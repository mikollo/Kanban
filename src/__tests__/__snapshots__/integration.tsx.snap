// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List can be added 1`] = `
<div>
  <div
    class="sc-bdVaJa jYKnyI"
  >
    <div
      class="sc-bxivhb hoWKiU"
    >
      <input
        class="sc-htpNat iHAoa"
        value="Some list"
      />
      <div
        class="sc-EHOje iutMph"
      >
        <input
          class="sc-bwzfXH dBYOUD"
          placeholder="Title"
        />
        <input
          class="sc-bwzfXH dBYOUD"
          placeholder="Description"
        />
        <button
          class="sc-ifAKCX jHzvVK"
        >
          Add new task
        </button>
      </div>
    </div>
    <div
      class="sc-bxivhb hoWKiU"
    >
      <input
        class="sc-bwzfXH dBYOUD"
        name="lol"
        placeholder="New list name"
        type="text"
      />
      <button
        class="sc-ifAKCX jHzvVK"
      >
        Add new list
      </button>
    </div>
  </div>
</div>
`;
